lib/elixir_analyzer/exercise_test/common_checks/unless_with_else.ex

```diff
@@ -0,0 +1,67 @@
+defmodule ElixirAnalyzer.ExerciseTest.CommonChecks.UnlessWithElse do
+  @moduledoc """
+  Reports when an unless clause is used followed by an else clause.
+
+  Common check to be run on every single solution.
+  """
+  alias ElixirAnalyzer.Constants
+  alias ElixirAnalyzer.Comment
+
+  @def_ops [:def, :defp, :defmacro, :defmacrop, :defguard, :defguardp]
+
+  @spec run(Macro.t()) :: [{:pass | :fail | :skip, %Comment{}}]
+  def run(ast) do
+    {_, issues} = Macro.prewalk(ast, [], &traverse/2)
+
+    # correct_implementation = if_else_implementation()
+    # wrong_implementation = issues
+
+    # if else_block? do
+    # [
+    #   {:fail,
+    #    %Comment{
+    #      type: :actionable,
+    #      comment: Constants.solution_unless_with_else(),
+    #      params: %{
+    #        expected: correct_implementation,
+    #        actual: wrong_implementation
+    #      }
+    #    }}
+    # ]
+    # end
+  end
+
+  defp traverse({:unless, meta, args} = ast, issues) do
+    if else_block?(args) do
+      a = {ast, issues}
+      IO.inspect(a, label: "a")
+    else
+      nil
+    end
+  end
+
+  defp else_block?(args) do
+    case else_block_for(args) do
+      {:ok, else_block} -> else_block
+      nil -> nil
+    end
+  end
+
+  defp else_block_for(args) when is_list(args) do
+    Enum.find_value(args, &find_keyword(&1, :else))
+  end
+
+  defp if_else_implementation({:unless, meta, args} = ast) do
+    [_, [do: do_block, else: else_block] = block] = args
+  end
```

Comment:
Here I intend to transform the unless/else implementation into a if/else one, so, my plan is to get the information in the else block and transform it into a if block, and transform the do_block into a else one, but I am not sure how can I do this. I would appreciate a little help here :)

Reply:
OOC, Why do you need to transform it? What will you need to do with the transformed ast? I think the better first step for this would be to just identify it, and report it and the line it is on to the user as something to fix.  To fix and correct the `unless..else`into an `if..else` naively might not be straight forward and might lead to recommending bad practices.

For ex) What if a student wrote:
```elixir
unless not x do
  :a
else
  :b
end
```

Unless handled carefully, this might be corrected to:

```elixir
if not not x do
  :b
else
  :a
end
```

---

As a second step I think it would be really cool to transform it and then be able to return it to the user, but I think we need to identify common situations to avoid the above.  To answer the original question though, to reassemble the ast you just need to return it in the right format:

```elixir
  defp if_else_implementation({:unless, meta, args}) do
    [condition, [do: do_block, else: else_block]] = args
    inverted_condition = {:not, [context: Elixir, import: Kernel], [condition]}

    {:if, meta, [inverted_condition, [do: else_block, else: do_block]]}
  end
```

Reply:
Transforming `unless/else` to `if/else` should be straight-forward because we don't need to touch the condition:
```
iex()> unless true, do: "false", else: "true"
"true"

iex()> if true, do: "true", else: "false"
"true"
```

However, I am against doing this (now or later) for two reasons:
- modified AST changed back to a string will have formatting that might drastically differ from what the student wrote in their code
- we have no idea how long the code blocks will be, vertically or horizontally, trying to use them in the analyzer comment might produce a comment that doesn't fit on the website and isn't very useful for explaining the issue with `unless/else`.

Let's just use a generic code example in the analyzer comment that will not have anything to do with the student's solution, but will show very well the problem 🙂

Reply:
@angelikatyborska @neenjaw yeah, I didn't realize that tranforming the code to an if-else implementation would be problematic, but the points you guys raised really make sense.